<!DOCTYPE html>
<html lang="en">
<head>
	<title>Introduction to Snakemake</title>
	<meta charset="UTF-8">
	<meta http-equiv="x-ua-compatible" content="ie=edge">
	<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
	<!-- Twitter Card -->
	<!-- Ref: https://dev.twitter.com/cards/types/summary-large-image -->
	<!--
	<meta name="twitter:card" content="summary_large_image">
	<meta name="twitter:site" content="@ccwang002">
	<meta name="twitter:creator" content="@ccwang002">
	<meta name="twitter:title" content="Genomics in R">
	<meta name="twitter:description" content="An intrdouction of using R/Bioconductor to explore genomic annotaions, covering usage of OrgDB, TxDb, OrganismDb, BSgenome, and GenomicRanges.">
	<meta name="twitter:image" content="https://blog.liang2.tw/2016Talk-Genomics-in-R/pics/modPlot_GAPDH.png">
	-->
	<!-- Icon -->
	<link href="pics/snakemake_logo.svg" rel="icon" type="image/x-icon" />
	<!-- CSS Stle -->
	<link rel="stylesheet" href="lib/shower/themes/styles/screen-16x10.css">
	<link rel="stylesheet" href="lib/highlight/styles/github.css" type="text/css"/>
		<style>
			.caption p {
				font-size: 0.6em;
			}
			.slide.mid h2 {
				font-size: 110px;
				font-family: Roboto Light;
				font-weight: 300;
				color: white;
				line-height: 1.25em;
				text-shadow: 1px 1px 3px black;
				top: 36%;
				position: absolute;
			}
			.citation pre {
				text-align: left;
			}
			pre.medium {
				font-size: 0.75em;
			}
			pre.medium code {
				line-height: 1.5em;
			}
			pre.small {
				font-size: 0.5em;
			}
			pre.small code {
				line-height: 1.5em;
			}

			.sidenote {
				border: 2px dashed #9e9e9e;
				position: absolute;
				padding: 1em;
				line-height: 1.5em;
				color: #9e9e9e;
			}
			.sidenote p {
				margin-bottom: 0px;
			}

		</style>
</head>
<body class="shower list">
	<!-- Header in overview -->
	<header class="caption">
		<h1>Introduction to Snakemake</h1>
		<p><a href="http://liang2.tw">Liang-Bo Wang (Bobo)</a>, 2017-09-05</p>
	</header>

	<!-- Cover Slide -->
	<section id="cover" class="slide no-number">
		<h2 id="talk-header">Introduction to Snakemake</h2>
		<p id="talk-author">
			By <a href="http://liang2.tw" target="_blank">Bobo (Liang<sup>2</sup>)</a> under CC 4.0 BY license
		</p>
		<p id="usage-instr">
			<kbd>Esc</kbd> to <a class="back-to-overview" href="javascript:Podium.keydown(27);">overview</a><br>
			<kbd>←</kbd> <kbd>→</kbd> to navigate
		</p>
		<img src="pics/cover.jpg" class="cover w" alt="">
		<style>
			#talk-header,
			#talk-author {
				left: 56px;
			}
			#talk-header {
				color: #EEE;
				text-align: left;
				font-size: 100px;
				line-height: 1.3em;
				text-shadow: -1px 1px 3px #000;
				bottom: 160px;
				position: absolute;

			}
			#talk-author {
				line-height: 1.2em;
				text-shadow: -1px 1px 3px #000;
				left: 65px;
				bottom: 120px;
				position: absolute;
			}
			#talk-author a,
			#cover .back-to-overview {
				color: #EDDC1C;
			}
			#cover p {
				margin: 10px 0 0;
				color: #FFF;
				font-size: 32px;
				opacity: 1;
			}
			#usage-instr {
				position: absolute;
				text-align: right;
				right: 30px;
				bottom: 40px;
			}
			#usage-instr kbd {
				opacity: 0.8;
				color: #666;
				background-color: white;
			}
			#cover img {
				opacity: 0.95;
			}
			#cover {
				background: black;
			}
			#cover .back-to-overview {
				color: #EDDC1C;
				text-decoration: inherit; /* no underline */
			}
		</style>
	</section>

	<!-- Section: Intro -->
	<section id="what-is-snakemake" class="slide twocol l80">
		<div class="left">
			<h2>What is Snakemake?</h2>
			<p><a href="https://snakemake.readthedocs.io/en/stable/">Snakemake</a> is a Python-based workflow management tool.</p>
			<ul>
				<li>A workflow contains a set of <strong>rules</strong></li>
				<li>A rule describes how to creat output files/patterns<br>from input files/pattern</li>
				<li>The workflow file runs like a Python script</li>
				<ul>
					<li>Can import external Python libraries</li>
					<li>Can write one's own functions</li>
				</ul>
			</ul>
		</div>
		<div class="right">
			<img src="pics/snakemake_logo.svg" alt="">
		</div>
	</section>

	<section id="how-snakemake-runs" class="slide">
		<h2>How does Snakemake work?</h2>
		<ol>
			<li>Given the final output files, snakemake will find the rule that generates them</li>
			<li>The rule will specify the required input files</li>
			<li>If the input files exist, then we are good to run the commands</li>
			<li>If some input does not exist but there are rules to generate them, use those rules to find the required input files</li>
			<li>Repeat the process until all outputs can be generated from existing files</li>
			<li>Snakemake will start running the corresponding commands</li>
		</ol>
	</section>

	<section id="snakemake-in-action" class="slide">
		<h2>Run snakemake in action</h2>
		<p>Write RNA-seq pipeline in Snakemake to assess transcript abundance:</p>
		<ol>
			<li>Build the alignment index</li>
			<li>Map sample reads to genome</li>
			<li>Assess sample's transcript abundance</li>
		</ol>
		<p>Tools to use: HISAT2 + StringTie (<a href="https://www.nature.com/nprot/journal/v11/n9/full/nprot.2016.095.html">new Tuxedo pipeline</a>)</p>
	</section>

	<!-- Section: Setup -->
	<section id="setup-cover" class="slide">
		<h2 class="shout">Setup</h2>
	</section>

	<section id="setup-overview" class="slide">
		<h2>Setup the environment</h2>
		<p>We will use Python 3.6 and install all the tools using <a href="https://bioconda.github.io/">bioconda</a>.<br>
			Before continue, make sure one has a working:</p>
		<ul>
			<li><strong><a href="https://conda.io/miniconda.html">Conda</a></strong> (any miniconda or anaconda 2/3)</li>
			<li><strong><a href="https://www.docker.com/">Docker</a></strong></li>
			<li>Access to <strong><a href="https://console.cloud.google.com/">Google Cloud Console</a></strong> and create VM instances.</li>
		</ul>
	</section>

	<section id="setup-dir" class="slide">
		<h2>Working directory and example data</h2>
		<p>Everything will be under <code>~/snakemake_example</code></p>
		<p>Example data from <a href="https://github.com/griffithlab/rnaseq_tutorial/">Griffith Lab’s RNA-seq tutorial</a>: 2 conditions, 3 repeats.</p>
		<pre class="language-bash medium"><code>cd ~/snakemake_example
wget https://storage.googleapis.com/lbwang-playground/snakemake_rnaseq/griffithlab_brain_vs_uhr.tar.gz
tar xf griffithlab_brain_vs_uhr.tar.gz</code></pre>
	</section>

	<section id="setup-bioconda" class="slide">
		<h2>Setup Bioconda</h2>
		<p><a href="https://bioconda.github.io/">Bioconda</a> is a conda channel providing many bioinfomatics software.</p>
		<pre class="language-bash"><code>conda config --add channels defaults
conda config --add channels conda-forge
conda config --add channels bioconda</code>
		</pre>
	</section>

	<section id="setup-conda-env" class="slide">
		<h2>Create a conda env</h2>
		<p>Conda env can isolate all the tools so one can run multiple pipelines without worrying about version conflict of a same tool.<br>
			Create an env called <code>new-tuxedo</code> to install all the required tools for this pipeline.</p>
		<pre class="language-shell"><code>conda create -n new_tuxedo \
    python=3.6 snakemake hisat2 stringtie samtools</code></pre>
	</section>

	<section id="setup-conda-env-contd" class="slide">
		<h2>Use the conda env</h2>
		<pre class="language-bash"><code>$ source activate new_tuxedo       # Use the conda env
(new_tuxedo) $ hisat2 --version
~/miniconda3/envs/new_tuxedo/bin/hisat2-align-s version 2.1.0
(new_tuxedo) $ deactivate           # Exit the env

# The tool is not available outside
$ hisat2 --version
bash: hisat2: command not found</code></pre>
	</section>

	<!-- Section: Write the pipeline -->
	<section id="pipeline-cover" class="slide">
		<h2 class="shout">Pipeline<br>(Write Snakefile)</h2>
	</section>

	<section id="pipeline-overview" class="slide">
		<h2>Pipeline overview</h2>
		<ol>
			<li><strong>Build HISAT2 genome reference index for alignment (we are here)</strong></li>
			<li>Align sample reads to the genome by HISAT2</li>
			<li>Assess transcript by StringTie</li>
		</ol>
		<p>The pipeline will be written to a file <code>Snakefile</code></p>
	</section>

	<section id="hisat-cmd" class="slide">
		<h2>How to build HISAT2 index?</h2>
<pre class="language-bash"><code>hisat2_extract_splice_sites.py genes.gtf > <mark class="important">genes.ss</mark>
hisat2_extract_exons.py genes.gtf > <mark>genes.exon</mark>
hisat2-build -p 4 \
	--ss <mark class="important">genes.ss</mark> --exon <mark>genes.exon</mark> \
	index_prefix</code></pre>
		<ul>
			<li>Index files will be: index_prefix.1.ht2, ..., index_prefix.8.ht2</li>
			<li>Last comand depends on the output of the first two commands</li>
		</ul>
	</section>

	<section id="pipeline-genome-index" class="slide">
		<h2>Snakemake rules for the first two commands</h2>
		<pre lang="python" class="medium"><code>rule extract_genome_splice_sites:
    input: "griffithlab_brain_vs_uhr/GRCh38_Ens87_chr22_ERCC/genes_chr22_ERCC92.gtf"
    output: "hisat2_index/chr22_ERCC92.ss"
	shell: "hisat2_extract_splice_sites.py {input} > {output}"

rule extract_genome_exons:
    input: "griffithlab_brain_vs_uhr/GRCh38_Ens87_chr22_ERCC/genes_chr22_ERCC92.gtf"
    output: "hisat2_index/chr22_ERCC92.exon"
	shell: "hisat2_extract_exons.py {input} > {output}"</code></pre>
	</section>

	<section id="run-snakemake" class="slide">
		<h2>Try run snakemake</h2>
		<pre lang="text"><code>snakemake [OPTIONS] [RULE or OUTPUT]
  -n --dryrun   Don't actually run the pipeline
  -p            Print the command of each job

snakemake hisat2_index/chr22_ERCC92.ss
rm hisat2_index/chr22_ERCC92.ss</code></pre>
	</section>

	<section id="pipeline-genome-index-contd" class="slide tighter">
		<h2>For the third command</h2>
<pre lang="python" class="medium"><code>rule build_hisat_index:
    input:
        genome_fa="griffithlab_brain_vs_uhr/GRCh38_Ens87_chr22_ERCC/chr22_ERCC92.fa",
        splice_sites="hisat2_index/chr22_ERCC92.ss",
        exons="hisat2_index/chr22_ERCC92.exon",
    output: expand("hisat2_index/chr22_ERCC92.{ix}.ht2", ix=range(1, 9))
    shell:
        "hisat2-build -p 4 {input.genome_fa} "
        "--ss {input.splice_sites} --exon {input.exons} "
        "hisat2_index/chr22_ERCC92"</code></pre>
<p><code>expand()</code> is a Snakemake utility function. Run snakemake again by:</p>
<pre lang="bash"><code>snakemake -p build_hisat_index</code></pre>
	</section>

	<section id="pipeline-overview-2" class="slide">
		<h2>Move on to alignment</h2>
		<ol>
			<li>Build HISAT2 genome reference index for alignment</li>
			<li><strong>Align sample reads to the genome by HISAT2 (we are here)</strong></li>
			<li>Assess transcript by StringTie</li>
		</ol>
	</section>

	<section id="hisat-align-cmd" class="slide">
		<h2>How to use HISAT2 for alignment?</h2>
		<p>For each sample, supply their pair-end FASTQs and index, and pipe the output to samtools.
			Output will be a BAM file per sample.</p>
<pre class="language-text"><code>hisat2 -p 4 --dta -x index_prefix \
    -1 sample_r1.fastq -2 sample_r2.fastq | \
	samtools sort -@ 4 -o sample.bam</code></pre>
		<p>How to apply a rule for different samples?</p>
	</section>

	<section id="pipeline-align" class="slide">
		<pre class="language-python medium">
<code>rule align_hisat:
    input:
        fq1="griffithlab_brain_vs_uhr/HBR_UHR_ERCC_ds_10pc/{sample}.read1.fastq.gz",
        fq2="griffithlab_brain_vs_uhr/HBR_UHR_ERCC_ds_10pc/{sample}.read2.fastq.gz",
        hisat2_index=expand("hisat2_index/chr22_ERCC92.{ix}.ht2", ix=range(1, 9)),
    output: "align_hisat2/{sample}.bam"
    threads: 4
    shell:
        "hisat2 -p {threads} --dta -x hisat2_index/chr22_ERCC92 "
        "-1 {input.fq1} -2 {input.fq2} | "
		"samtools sort -@ {threads} -o {output}"</code>
		</pre>
		<ul>
			<li><code>threads</code> entry specifies how many CPUs the rule will use.</li>
			<li><code>{sample}</code> is a wildcard so similar outputs can share a rule.</li>
		</li>
	</section>

	<section id="wildard" class="slide tighter">
		<h2>More on wildcard</h2>
		<pre class="language-python medium">
<code>rule simple:
    input:
        fq1=".../{sample}.read1.fastq.gz",
        fq2=".../{sample}.read2.fastq.gz",
    output: "out/{sample}.bam"
    threads: 4
    shell:
		"mycmd process {input.fq1} {input.fq2} > {output}"</code>
		</pre>
		<p>While run <code>snakemake out/A.bam</code>, it matches the output pattern and knows <code>sample = A</code> and replace all the wildcard</p>
		<p>But how to apply to all samples? or to get all the name of the samples?</p>
	</section>

	<section id="glob" class="slide">
		<h2>To get all the samples' name</h2>
		<pre class="language-python medium">
<code>SAMPLES, *_ = glob_wildcards(
    'griffithlab_brain_vs_uhr/HBR_UHR_ERCC_ds_10pc/'
	'{sample}.read1.fastq.gz')

rule align_all_samples:
    input: expand("align_hisat2/{sample}.bam", sample=SAMPLES)</code>
		</pre>
		<p class="note">One can always hard code <code>SAMPLES = ['A', 'B', ...]</code></p>
		<pre lang="bash"><code>snakemake -j 8 -p align_all_samples</code></pre>
	</section>

	<section id="dag" class="slide">
		<h2>How Snakemake resolves jobs</h2>
		<p>When Snakemake is figuring out what input are required to generate the desired output, a DAG is created along the way.</p>
		<div class="citation">
			<img src="pics/align_dag.svg" class="fill w" alt="">
		</div>
		<p>Each node in the DAG represent a job. Job that has its dependency ready can be run in parallel, say, no connected edges.</p>
	</section>

	<section id="pipeline-stringtie" class="slide">
		<h2>Move on to transcript assessment</h2>
		<ol>
			<li>Build HISAT2 genome reference index for alignment</li>
			<li>Align sample reads to the genome by HISAT2</li>
			<li><strong>Assess transcript by StringTie (we are here)</strong></li>
		</ol>
		<p>Download <a href="https://gist.github.com/ccwang002/2659b19439b6205284c0ae68ca06345d">the full Snakefile</a> to save some keystrokes.</p>
		<p>Run the full pipeline by <code>snakemake -j 8 -p quant_all_samples</code></p>
	</section>

	<section id="more-features" class="slide">
		<h2>More features of Snakemake</h2>
		<ul>
			<li><a href="https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#snakefiles-input-functions">Use function as a intput</a></li>
			<li><code>log:</code> to <a href="https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#log-files">store log files</a></li>
			<li><a href="https://snakemake.readthedocs.io/en/stable/snakefiles/configuration.html">Config and consants</a></li>
			<li>Use <code>run:</code> instead of <code>shell:</code> for running custom Python snippet</li>
			<li><a href="https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#external-scripts">Use external R/Python script instead of shell command</a></li>
		</ul>
	</section>

	<!-- Section: Docker -->
	<section id="docker-cover" class="slide">
		<h2 class="shout">Docker</h2>
	</section>

	<section id="sakemake-in-docker" class="slide">
		<h2>Run Snakemake in Docker</h2>
		<ul>
			<li>We need to pack Snakemake and all the tools inside a Docker image</li>
			<li>Map the Snakefile and working directory inside Docker</li>
			<li>Rest of the setup is the same</li>
		</ul>
	</section>

	<section id="docker-image" class="slide">
		<h2>Make a Docker image</h2>
		<p>Use <a href="https://hub.docker.com/r/lbwang/snakemake-conda-rnaseq/">my Docker image</a> or a minimal working <code>Dockerfile</code>:</p>
		<pre class="language-docker"><code>FROM continuumio/miniconda3
RUN conda config --add channels conda-forge \
    && conda config --add channels bioconda \
    && conda install -y python=3.6 nomkl \
        stringtie samtools hisat2 snakemake \
        google-cloud-storage</code></pre>
	</section>

	<section id="run-snakemake-in-docker" class="slide tighter">
		<pre class="language-bash"><code>cd ~/snakemake_example
docker run -t                       \
    -v $(pwd):/analysis             \
    lbwang/snakemake-conda-rnaseq   \
    snakemake -j 2 --timestamp      \
        -s /analysis/Snakefile --directory /analysis \
        quant_all_samples</code></pre>
			<ul>
				<li><code>-v ...</code> maps the current directory to /analysis inside Docker</li>
				<li><code>-s ...</code> lets Snakemake find the mapped Snakefile<br>
					<code>--directory</code> ...  and working dir</li>
			</ul>
	</section>

	<!-- Section: Google Cloud -->
	<section id="gcloud-cover" class="slide">
		<h2 class="shout">Google Cloud</h2>
	</section>

	<section id="gcloud-setup" class="slide">
		<h2>Google Cloud Setup</h2>
		<ul>
			<li>A writable Google Cloud Storage <strong>(GS) bucket/path</strong><br>
				Say, <code>gs://dinglab/lbwang/snakemake_demo/</code></li>
			<li>A Google Cloud Compute <strong>(GCE) VM instance</strong></li>
			<ul>
				<li>10GB disk space</li>
				<li>Has write access to that GS bucket/path</li>
				<li><strong>Docker</strong> installed</li>
			</ul>
			<li>An empty working directory at <code>~/snakemake_example</code></li>
		</ul>
	</section>

	<section id="gcloud-snakemake" class="slide">
		<h2>Let Snakemake use remote files</h2>
		<pre class="language-python">
<code>from snakemake.remote.GS import RemoteProvider \
	as GSRemoteProvider
GS = GSRemoteProvider()

rule xxx:
    input: GS.remote("readonlybucket/path/to/file")
	output: GS.remote("writablebucket/path/to/file")</code>
		</pre>
	</section>

	<section id="gcloud-remote-input" class="slide">
		<h2>Make all the input files on GS</h2>
		<p>I've uploaded a copy on Google Cloud Storage already at<br>
			<code>gs://lbwang-playground/snakemake_rnaseq</code>.</p>
		<pre class="language-python medium">
<code>rule align_hisat:
    input:
        fastq1=GS.remote("lbwang-playground/.../{sample}.read1.fastq.gz"),
        fastq2=GS.remote("lbwang-playground/.../{sample}.read2.fastq.gz")
    output: "align_hisat2/{sample}.bam"
    shell: ...</code>
		</pre>
		<p>Only input/output entries are affected. One can update all paths to be on GS.</p>
	</section>

	<section id="gcloud-input-bucket" class="slide">
		<h2>All input files on my readonly bucket</h2>
		<pre class="language-bash small">
<code><mark class="important">gsutil ls -r gs://lbwang-playground/snakemake_rnaseq/griffithlab_brain_vs_uhr/</mark>
gs://lbwang-playground/snakemake_rnaseq/griffithlab_brain_vs_uhr/:

gs://lbwang-playground/snakemake_rnaseq/griffithlab_brain_vs_uhr/GRCh38_Ens87_chr22_ERCC/:
gs://lbwang-playground/snakemake_rnaseq/griffithlab_brain_vs_uhr/GRCh38_Ens87_chr22_ERCC/chr22_ERCC92.fa
gs://lbwang-playground/snakemake_rnaseq/griffithlab_brain_vs_uhr/GRCh38_Ens87_chr22_ERCC/genes_chr22_ERCC92.gtf

gs://lbwang-playground/snakemake_rnaseq/griffithlab_brain_vs_uhr/HBR_UHR_ERCC_ds_10pc/:
gs://lbwang-playground/snakemake_rnaseq/griffithlab_brain_vs_uhr/HBR_UHR_ERCC_ds_10pc/HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz
gs://lbwang-playground/snakemake_rnaseq/griffithlab_brain_vs_uhr/HBR_UHR_ERCC_ds_10pc/HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz
gs://lbwang-playground/snakemake_rnaseq/griffithlab_brain_vs_uhr/HBR_UHR_ERCC_ds_10pc/HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz
gs://lbwang-playground/snakemake_rnaseq/griffithlab_brain_vs_uhr/HBR_UHR_ERCC_ds_10pc/HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz
gs://lbwang-playground/snakemake_rnaseq/griffithlab_brain_vs_uhr/HBR_UHR_ERCC_ds_10pc/HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz
gs://lbwang-playground/snakemake_rnaseq/griffithlab_brain_vs_uhr/HBR_UHR_ERCC_ds_10pc/HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz
gs://lbwang-playground/snakemake_rnaseq/griffithlab_brain_vs_uhr/HBR_UHR_ERCC_ds_10pc/UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz
gs://lbwang-playground/snakemake_rnaseq/griffithlab_brain_vs_uhr/HBR_UHR_ERCC_ds_10pc/UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz
gs://lbwang-playground/snakemake_rnaseq/griffithlab_brain_vs_uhr/HBR_UHR_ERCC_ds_10pc/UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz
gs://lbwang-playground/snakemake_rnaseq/griffithlab_brain_vs_uhr/HBR_UHR_ERCC_ds_10pc/UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz
gs://lbwang-playground/snakemake_rnaseq/griffithlab_brain_vs_uhr/HBR_UHR_ERCC_ds_10pc/UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz
gs://lbwang-playground/snakemake_rnaseq/griffithlab_brain_vs_uhr/HBR_UHR_ERCC_ds_10pc/UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz</code>
		</pre>
	</section>

	<section id="snakemake-default-remote" class="slide">
		<h2>Point all the paths under GS</h2>
		<p>Snakemake allow user to pass <code>--default-remote-provider</code> and <code>--default-remote-prefix</code> so all the local path will be on GS.</p>
		<ul>
			<li>For example, when the default remote prefix being<br>
				<code>dinglab/lbwang/snakemake_demo</code></li>
			<li>Then the output path <code>align_hisat2/{sample}.bam</code> becomes<br>
				<code>dinglab/lbwang/snakemake_demo/align_hisat2/{sample}.bam</code></li>
		</ul>
	</section>

	<section id="remote-snakefile" class="slide">
		<h2>Modify our Snakefile</h2>
		<p>The modified Snakefile can be found <a href="https://gist.github.com/ccwang002/2686840e90574a67a673ec4b48e9f036">here</a>. Changes worth noting:</p>
		<ul>
			<li>Host all the input source files on GS using <code>GS.remote(...)</code></li>
			<li>Modify <code>FULL_HISAT2_INDEX_PREFIX</code> to be <code>{WRITABLE_BUCKET_PATH}/hisat2_index/chr22_ERCC92</code></li>
			<ul>
				<li>A special case of path changing because HISAT2 command doesn't accept the path to the indices but just the prefix.</li>
			</ul>
		</ul>
	</section>

	<section id="snakemake-in-docker-on-gcloud" class="slide">
<pre class="language-bash"><code>docker run -t -i                           \
    -v $(pwd):/analysis                         \
    -v ~/.config/gcloud:/root/.config/gcloud    \
    lbwang/snakemake-conda-rnaseq               \
    snakemake -j 4 --timestamp --verbose -p --keep-remote   \
        -s /analysis/Snakefile --directory /analysis        \
		--default-remote-provider GS                        \
		--default-remote-prefix "{WRITABLE_BUCKET_PATH}"    \
        quant_all_samples</code></pre>
	</section>

	<section id="snakemake-demo" class="slide black">
		<video class="cover fill w" controls preload="metadata">
		<source src="pics/pipeline_demo_6x.mp4" type="video/mp4">
		Your browser doesn't support HTML5 video in MP4 with H.264.
		You can still download the <a href="pics/pipeline_demo_6x.mp4">screencast</a> and view it locally.
		</video>
	</section>

	<section id="how-snakemake-deals-with-remote-files" class="slide tighter">
		<pre class="language-text">
<code>~/snakemake_example
├── lbwang-playground/
│   └── snakemake_rnaseq/
│       └── griffithlab_brain_vs_uhr/
│           ├── GRCh38_Ens87_chr22_ERCC/
│           └── HBR_UHR_ERCC_ds_10pc/
├── {WRITABLE_BUCKET_PATH}/
│   ├── align_hisat2/
│   ├── hisat2_index/
│   └── stringtie/
└── Snakefile</code>
		</pre>
	</section>

	<!-- Section: Conclusion -->
	<section id="futher-reading" class="slide">
		<h2>Futher reading</h2>
		<ul>
			<li><a href="https://snakemake.readthedocs.io/en/stable/index.html">Snakemake's official documentation</a></li>
			<li><a href="http://slides.com/johanneskoester/snakemake-broad-2015">High-level introduction of Snakemake</a></li>
			<li><a href="https://blog.liang2.tw/posts/2017/08/snakemake-google-cloud/">My blog post covering the same pipeline in detail</a></li>
		</ul>
	</section>

	<!-- Section: About me -->
	<section id="about-me" class="slide no-number twocol l60">
		<div class="left">
			<h2>About Me</h2>
			<ul>
				<li>Liang-Bo Wang, aka Bobo or Liang<sup>2</sup></li>
				<li>Computational and Systems Biology PhD student in <a href="http://dinglab.wustl.edu/">Ding Lab at WUSTL</a></li>
				<li>Speak Python and R</li>
			</ul>
		</div>
		<div class="right">
			<img class="fill w" src="pics/me.jpg" alt="">
		</div>
	</section>

	<!-- End Slide -->
	<section id='end' class='slide no-number'>
		<h2>Q&A</h2>
		<img src="pics/end.jpg" class="cover w" alt="">
		<style>
			#end h2 {
				position: absolute;
				left: 12%;
				top: 17%;
				color: rgb(98, 175, 212);
				text-align: right;
				text-shadow: none;
				font-size: 150px;
			}
		</style>
	</section>

	<!-- END OF SLIDE CONTENT -->
	<!--
	To hide progress bar from entire presentation
	just remove “progress” element.
	-->
	<div class="progress"></div>
	<footer class="badge">
		<a href="https://github.com/ccwang002/2017Talk-Snakemake">
			<svg viewBox="0 0 81 79" aria-hidden="true">
				<path d="M40.5 0a40.5 40.5 0 0 0-12.8 78.93c2 .37 2.76-.88 2.76-2s0-3.51-.05-6.89c-11.27 2.45-13.64-5.43-13.64-5.43-1.84-4.68-4.5-5.92-4.5-5.92-3.68-2.51.28-2.46.28-2.46 4.06.29 6.2 4.17 6.2 4.17 3.61 6.19 9.48 4.4 11.79 3.37a8.65 8.65 0 0 1 2.57-5.41c-9-1-18.45-4.5-18.45-20a15.65 15.65 0 0 1 4.17-10.87 14.57 14.57 0 0 1 .4-10.72s3.4-1.09 11.14 4.15a38.39 38.39 0 0 1 20.28 0c7.73-5.24 11.13-4.15 11.13-4.15a14.55 14.55 0 0 1 .4 10.72 15.63 15.63 0 0 1 4.16 10.87c0 15.56-9.47 19-18.49 20 1.45 1.25 2.75 3.72 2.75 7.5v11.11c0 1.08.73 2.34 2.78 1.95A40.51 40.51 0 0 0 40.5 0z"/>
			</svg>
			Sources<br>on GitHub
		</a>
	</footer>

	<!-- Library -->
	<script src="lib/shower/shower.min.js"></script>
	<script src="lib/highlight/highlight.pack.js" type="text/javascript" charset="utf-8"></script>
	<script>
		hljs.initHighlightingOnLoad();
	</script>
	<script>
		// Emulate Esc key for special links so on mobile user can go back to list view
		// A long term solution is to wait for Shower.js to expose its API
		// Simulate Esc key solution from SO: https://stackoverflow.com/a/10520017
		Podium = {};
		Podium.keydown = function(k) {
			var oEvent = document.createEvent('KeyboardEvent');

			// Chromium Hack
			Object.defineProperty(oEvent, 'keyCode', {
						get : function() {
							return this.keyCodeVal;
						}
			});
			Object.defineProperty(oEvent, 'which', {
						get : function() {
							return this.keyCodeVal;
						}
			});

			if (oEvent.initKeyboardEvent) {
				oEvent.initKeyboardEvent("keydown", true, true, document.defaultView, false, false, false, false, k, k);
			} else {
				oEvent.initKeyEvent("keydown", true, true, document.defaultView, false, false, false, false, k, 0);
			}

			oEvent.keyCodeVal = k;

			if (oEvent.keyCode !== k) {
				alert("keyCode mismatch " + oEvent.keyCode + "(" + oEvent.which + ")");
			}

			document.dispatchEvent(oEvent);
		}
	</script>
	<!-- Mathjax -->
	<!-- online Mathjax CDN -->
	<!--<script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>-->
</body>
</html>
